File: coco/src/test/java/coco/ep/M6Test.java
package coco.ep;

import coco.ep.m5.Exp;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class M6Test {

    public interface TestTemplate<FT> extends coco.ep.m6.Factory<FT>, M5Test.TestTemplate<FT> {
        default void test() {
            M5Test.TestTemplate.super.test();

            coco.ep.m5.Exp<FT> lit1 = convert(this.lit(1.0));
            coco.ep.m5.Exp<FT> lit2 = convert(this.lit(2.0));
            coco.ep.m5.Exp<FT> mult1 = convert(this.mult(lit1, lit2));
            coco.ep.m5.Exp<FT> divd1 = convert(this.divd(lit1, lit2));
            coco.ep.m5.Exp<FT> add1 = convert(this.add(lit1, lit2));
            coco.ep.m5.Exp<FT> sub1 = convert(this.sub(lit1, lit2));
            Exp<FT> neg1 = convert(this.neg(lit1));

            Assertions.assertTrue(convert(lit1).equals(convert(lit1)));
            Assertions.assertFalse(convert(lit1).equals(convert(lit2)));
            Assertions.assertTrue(convert(mult1).equals(convert(mult1)));
            Assertions.assertFalse(convert(mult1).equals(convert(divd1)));
            Assertions.assertTrue(convert(divd1).equals(convert(divd1)));
            Assertions.assertFalse(convert(divd1).equals(convert(add1)));
            Assertions.assertTrue(convert(add1).equals(convert(add1)));
            Assertions.assertFalse(convert(add1).equals(convert(sub1)));
            Assertions.assertTrue(convert(sub1).equals(convert(sub1)));
            Assertions.assertFalse(convert(sub1).equals(convert(neg1)));
            Assertions.assertTrue(convert(neg1).equals(convert(neg1)));
            Assertions.assertFalse(convert(neg1).equals(convert(mult1)));

            Assertions.assertTrue(convert(lit1).eql(convert(lit1)));
            Assertions.assertFalse(convert(lit1).eql(convert(lit2)));
            Assertions.assertFalse(convert(mult1).eql(convert(lit2)));
            Assertions.assertTrue(convert(mult1).eql(convert(mult1)));
            Assertions.assertFalse(convert(mult1).eql(convert(divd1)));
            Assertions.assertTrue(convert(divd1).eql(convert(divd1)));
            Assertions.assertFalse(convert(divd1).eql(convert(add1)));
            Assertions.assertTrue(convert(add1).eql(convert(add1)));
            Assertions.assertFalse(convert(add1).eql(convert(sub1)));
            Assertions.assertTrue(convert(sub1).eql(convert(sub1)));
            Assertions.assertFalse(convert(sub1).eql(convert(neg1)));
            Assertions.assertTrue(convert(neg1).eql(convert(neg1)));
            Assertions.assertFalse(convert(neg1).eql(convert(mult1)));

            Assertions.assertTrue(convert(this.sub(this.lit(1.0), this.lit(73.0))).equals(this.sub(this.lit(1.0), this.lit(73.0))));
            Assertions.assertFalse(convert(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0))).equals(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(3.0))));
            Assertions.assertTrue(convert(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0))).equals(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0))));
            Assertions.assertTrue(convert(this.neg(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0)))).equals(this.neg(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0)))));
            Assertions.assertFalse(convert(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0))).equals(this.neg(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0)))));
            Assertions.assertFalse(convert(this.divd(this.lit(6.0), this.lit(2.0))).equals(this.divd(this.lit(8.0), this.lit(2.0))));
            Assertions.assertTrue(convert(this.divd(this.lit(6.0), this.lit(2.0))).equals(this.divd(this.lit(6.0), this.lit(2.0))));
            Assertions.assertTrue(convert(this.add(this.lit(5.0), this.lit(3.0))).equals(this.add(this.lit(5.0), this.lit(3.0))));
            Assertions.assertFalse(convert(this.add(this.lit(5.0), this.lit(3.0))).equals(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(3.0))));
        }
    }
    private static class ActualTest implements M6Test.TestTemplate<coco.ep.m6.finalized.Exp>, coco.ep.m6.finalized.Factory {}

    @Test
    public void testTest() { new M6Test.ActualTest().test(); }
}

File: coco_scala/src/test/scala/exp/m6/M6Test.scala
package exp.m6

import org.scalatest.FunSuite

trait TestTemplate[FT] extends Factory[FT] with exp.m4.TestTemplate[FT] {
  val suite:FunSuite
  import suite._

  override def test() : Unit = {
    super.test()


    val lit1 = this.lit(1.0)
    val lit2 = this.lit(2.0)
    val mult1 = this.mult(lit1, lit2)
    val divd1 = this.divd(lit1, lit2)
    val add1 = this.add(lit1, lit2)
    val sub1 = this.sub(lit1, lit2)
    val neg1 = this.neg(lit1)

    assert(lit1.equals(lit1))
    assert(!lit1.equals(lit2))
    assert(mult1.equals(mult1))
    assert(!mult1.equals(divd1))
    assert(divd1.equals(divd1))
    assert(!divd1.equals(add1))
    assert(add1.equals(add1))
    assert(!add1.equals(sub1))
    assert(sub1.equals(sub1))
    assert(!sub1.equals(neg1))
    assert(neg1.equals(neg1))
    assert(!neg1.equals(mult1))

    assert(lit1.eql(lit1))
    assert(!lit1.eql(lit2))
    assert(mult1.eql(mult1))
    assert(!mult1.eql(divd1))
    assert(divd1.eql(divd1))
    assert(!divd1.eql(add1))
    assert(add1.eql(add1))
    assert(!add1.eql(sub1))
    assert(sub1.eql(sub1))
    assert(!sub1.eql(neg1))
    assert(neg1.eql(neg1))
    assert(!neg1.eql(mult1))

    assert(this.sub(this.lit(1.0), this.lit(73.0)).equals(this.sub(this.lit(1.0), this.lit(73.0))))
    assert(!this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0)).equals(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(3.0))))
    assert(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0)).equals(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0))))
    assert(this.neg(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0))).equals(this.neg(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0)))))
    assert(!this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0)).equals(this.neg(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0)))))
    assert(!this.divd(this.lit(6.0), this.lit(2.0)).equals(this.divd(this.lit(8.0), this.lit(2.0))))
    assert(this.divd(this.lit(6.0), this.lit(2.0)).equals(this.divd(this.lit(6.0), this.lit(2.0))))
    assert(this.add(this.lit(5.0), this.lit(3.0)).equals(this.add(this.lit(5.0), this.lit(3.0))))
    assert(!this.add(this.lit(5.0), this.lit(3.0)).equals(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(3.0))))
  }
}

class M6Test extends FunSuite { self =>

  object ActualTest extends TestTemplate[exp.m6.finalized.Exp] with finalized.Factory {
    val suite: FunSuite = self
  }

  test("M6") {
    ActualTest.test()
  }
}
